libmscore/tremolo.cpp

```diff
@@ -234,71 +235,38 @@ void Tremolo::layoutOneNoteTremolo(qreal x, qreal y, qreal _spatium)
-                     // UP
-                     {
-                        // even line   odd line
-                        { -3,          -3          },  // line 1
-                        { -2 - 2 * .8, -2 - 2 * .8 },  // line 2
-                        { -2 - 4 * .8, -2 - 4 * .8 },  // line 3
-                        { -2 - 6 * .8, -2 - 6 * .8 }   // line 4
-                        }
-                     },
-                  // stem with beam
-                  {
-                     // DOWN
-                     {
-                        // even line   odd line
-                        { 3,           3          },  // line 1
-                        { 2,           2          },  // line 2
-                        { 2,           2          },  // line 3
-                        { 2,           2          }   // line 4
-                        },
-                     // UP
-                     {
-                        // even line   odd line
-                        { -3,          -3          },  // line 1
-                        { -2 - 2 * .8, -2 - 2 * .8 },  // line 2
-                        { -2 - 4 * .8, -2 - 4 * .8 },  // line 3
-                        { -2 - 6 * .8, -2 - 6 * .8 }   // line 4
-                        }
-                     },
-                  };
-            int idx = _chord1->hook() ? 1 : (_chord1->beam() ? 2 : 0);
-            y = (line + t[idx][up][_lines-1][line & 1]) * .5 * _spatium;
+            const qreal td = score()->styleS(Sid::tremoloDistance).val();
+            const qreal sw = score()->styleS(Sid::tremoloStrokeWidth).val();
+
+            qreal t;
+            // nearest distance between note and tremolo stroke should be no less than 3
+            if (chord()->hook() || chord()->beam()) {
+                  t = up ? -3.0 - (2.0 * (lines() - 1)) * td - 2.0 * sw : 3.0;
+                  }
+            else { 
+                  if      (!up && !(line & 1)) // stem is down; even line
+                        t = qMax(6.0  - (2.0 * (lines() - 1)) * td - 2.0 * sw, 3.0);
+                  else if (!up &&  (line & 1)) // stem is down; odd line
+                        t = qMax(5.0  - (2.0 * (lines() - 1)) * td - 2.0 * sw, 3.0);
+                  else if ( up && !(line & 1)) // stem is up; even line
+                        t = qMin(-3.0 - (2.0 * (lines() - 1)) * td - 2.0 * sw, -6.0);
+                  else if ( up &&  (line & 1)) // stem is up; odd line
+                        t = qMin(-3.0 - (2.0 * (lines() - 1)) * td - 2.0 * sw, -5.0);
+                  }
+
+            qreal yLine = line + t;
+            // prevent stroke from going out of staff at the top while note is above staff, stem down
+            if (line < 0 && !chord()->up())
+                  yLine = qMax(yLine, 0.0);
+            // prevent stroke from going out of staff at the bottom while note is below staff, stem up
+            else if (line > ((staff()->lines(tick()) - 1) * 2) && chord()->up())
```

Comment:
The `line` conditional is not needed actually. Should I keep it for double-check?

Reply:
Deleted.